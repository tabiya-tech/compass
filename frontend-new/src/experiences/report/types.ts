import { Experience } from "src/experiences/experiencesDrawer/experienceService/experiences.types";

export type ReportProps = {
  name: string;
  email: string;
  phone: string;
  address: string;
  experiences: Experience[];
  conversationConductedAt: string | null;
};

export enum ExportFormat {
  PDF = "pdf",
  DOCX = "docx",
}

export interface IReportFormatProvider {
  download: () => void;
}
