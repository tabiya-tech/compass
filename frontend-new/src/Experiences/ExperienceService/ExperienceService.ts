import { getBackendUrl } from "src/envService";
import { getServiceErrorFactory } from "src/error/ServiceError/ServiceError";
import { fetchWithAuth } from "src/utils/fetchWithAuth/fetchWithAuth";
import { StatusCodes } from "http-status-codes";
import ErrorConstants from "src/error/ServiceError/ServiceError.constants";
import { Experience } from "src/Experiences/ExperienceService/Experiences.types";

export default class ExperienceService {
  readonly experiencesEndpointUrl: string;
  readonly apiServeUrl: string;
  private readonly sessionId: number;

  constructor(sessionId: number) {
    this.apiServeUrl = getBackendUrl();
    this.experiencesEndpointUrl = `${this.apiServeUrl}/conversation/experiences`;
    this.sessionId = sessionId;
  }

  static getInstance(sessionId: number): ExperienceService {
    return new ExperienceService(sessionId);
  }

  public getSessionId(): number {
    return this.sessionId;
  }

  async getExperiences(): Promise<Experience[]> {
    const constructedExperiencesUrl = `${this.experiencesEndpointUrl}?session_id=${this.getSessionId()}`;
    const errorFactory = getServiceErrorFactory(
      "ExperienceService",
      "getExperiences",
      "GET",
      constructedExperiencesUrl
    );

    let response: Response;
    let responseBody: string;
    response = await fetchWithAuth(constructedExperiencesUrl, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
      expectedStatusCode: StatusCodes.OK,
      serviceName: "ExperienceService",
      serviceFunction: "getExperiences",
      failureMessage: "Failed to retrieve experiences",
      expectedContentType: "application/json",
    });
    responseBody = await response.text();

    let experiencesResponse: Experience[];
    try {
      experiencesResponse = JSON.parse(responseBody);
    } catch (error: unknown) {
      throw errorFactory(
        response.status,
        ErrorConstants.ErrorCodes.INVALID_RESPONSE_BODY,
        "Response did not contain valid JSON",
        {
          responseBody,
          error: error,
        }
      );
    }

    return experiencesResponse;
  }
}
