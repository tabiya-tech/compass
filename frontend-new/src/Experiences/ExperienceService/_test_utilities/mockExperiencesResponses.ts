import { Experience, Skill, WorkType } from "src/Experiences/ExperienceService/Experiences.types";
import { faker } from "@faker-js/faker";
import { v4 as uuidv4 } from "uuid";
import { format } from "date-fns";

export const mockExperiences: Experience[] = [
  {
    UUID: "c3c8b43d-73dd-4c6d-9e71-010492e86d5e",
    start_date: "2022",
    end_date: "Present",
    experience_title: "Project Manager",
    company: "Business Inc",
    location: "New York, NY",
    work_type: WorkType.FORMAL_SECTOR_WAGED_EMPLOYMENT,
    top_skills: [
      {
        UUID: "2512652c-49ba-4751-b654-8fc525ce98ca",
        preferredLabel: "Management",
        description: "The process of dealing with or controlling things or people",
        altLabels: ["Leadership", "Administration"],
      },
      {
        UUID: "815f3660-4c7f-43ba-859d-cf50dd527fe0",
        preferredLabel: "Communication",
        description: "The imparting or exchanging of information or news",
        altLabels: ["Interpersonal Communication", "Public Speaking"],
      },
    ],
  },
  {
    UUID: "c3c8b43d-73dd-4c6d-9e71-010492e86d5e",
    start_date: "2022",
    end_date: "Present",
    experience_title: "Project Manager",
    company: "Business Inc",
    location: "New York, NY",
    work_type: WorkType.FORMAL_SECTOR_UNPAID_TRAINEE_WORK,
    top_skills: [
      {
        UUID: "2512652c-49ba-4751-b654-8fc525ce98ca",
        preferredLabel: "Management",
        description: "The process of dealing with or controlling things or people",
        altLabels: ["Leadership", "Administration"],
      },
      {
        UUID: "815f3660-4c7f-43ba-859d-cf50dd527fe0",
        preferredLabel: "Communication",
        description: "The imparting or exchanging of information or news",
        altLabels: ["Interpersonal Communication", "Public Speaking"],
      },
    ],
  },
];

const allWorkTypes = [
  WorkType.SELF_EMPLOYMENT,
  WorkType.FORMAL_SECTOR_WAGED_EMPLOYMENT,
  WorkType.FORMAL_SECTOR_UNPAID_TRAINEE_WORK,
  WorkType.UNSEEN_UNPAID,
];

const skillsLabels = [
  "work in an organised manner",
  "administer ingredients in food production",
  "sales activities",
  "unload supplies",
  "perform outdoor cleaning activities",
  "selling Kotas",
  "helping Brother with Car",
  "community Volunteering",
  "ensured the project was completed on time",
  "took care of the garden",
  "communication",
  "helped with the family business",
  "time management",
  "organise product display",
  "assist customers",
  "maintain work area cleanliness",
  "teamwork",
  "attention to details",
  "problem solving",
];

const generateRandomSkill = (usedLabels: string[]): Skill => {
  let randomLabel;
  do {
    randomLabel = skillsLabels[Math.floor(Math.random() * skillsLabels.length)];
  } while (usedLabels.includes(randomLabel));
  usedLabels.push(randomLabel);
  return {
    UUID: uuidv4(),
    preferredLabel: randomLabel,
    description: faker.hacker.phrase(),
    altLabels: [faker.hacker.verb(), faker.hacker.adjective()],
  };
};

const formatShortDate = (date: Date): string => format(date, "MMM yyyy");

const generateRandomDate = (): string => {
  return formatShortDate(faker.date.past());
};

const generateStandaloneDate = (): string => {
  const chance = Math.random();
  return chance < 0.5 ? "A long time ago" : "Since I was five";
};

const generateRandomExperience = (workType?: WorkType): Experience => {
  const useStandaloneDate = Math.random() < 0.5;
  const startDate = useStandaloneDate ? generateStandaloneDate() : generateRandomDate();
  const endDate = useStandaloneDate ? "" : Math.random() > 0.5 ? generateRandomDate() : "Present";
  const randomWorkType = workType || allWorkTypes[Math.floor(Math.random() * allWorkTypes.length)];
  const usedLabels: string[] = [];

  return {
    UUID: uuidv4(),
    start_date: startDate,
    end_date: endDate,
    experience_title: faker.person.jobTitle(),
    company: faker.company.name(),
    location: faker.location.city(),
    work_type: randomWorkType,
    top_skills: [
      generateRandomSkill(usedLabels),
      generateRandomSkill(usedLabels),
      generateRandomSkill(usedLabels),
      generateRandomSkill(usedLabels),
      generateRandomSkill(usedLabels),
    ],
  };
};

export const generateRandomExperiences = (count: number): Experience[] => {
  const experiences: Experience[] = [];

  // ensure at least one experience is present for each work type
  allWorkTypes.forEach((workType) => {
    experiences.push(generateRandomExperience(workType));
  });

  // generate the rest of the experiences
  for (let i = 0; i < count; i++) {
    experiences.push(generateRandomExperience());
  }
  return experiences;
};
