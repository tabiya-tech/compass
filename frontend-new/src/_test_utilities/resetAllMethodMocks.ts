// Utility function to reset all mocks on the instanceâ€™s methods.
// @ts-ignore
export function resetAllMethodMocks(instance) {
  // Get the prototype of the instance to list instance methods.
  const proto = Object.getPrototypeOf(instance);

  // Iterate over all property names except the constructor.
  Object.getOwnPropertyNames(proto)
    .filter(
      (prop) =>
        prop !== "constructor" &&
        typeof instance[prop] === "function",
    )
    .forEach((methodName) => {
      const method = instance[methodName];
      // Check if the method is a Jest mock (spyOn attaches the mock properties)
      if (method && typeof method.mockReset === "function") {
        method.mockReset();
      }
    });
}