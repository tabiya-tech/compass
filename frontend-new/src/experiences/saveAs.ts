export function saveAs(blob: Blob, filename: string): void {
  // For Internet Explorer only
  if ((window.navigator as any)?.msSaveOrOpenBlob) {
    (window.navigator as any).msSaveOrOpenBlob(blob, filename);
    return;
  }

  // For other browsers, create a temporary URL
  const url = URL.createObjectURL(blob);

  // Create a temporary anchor element
  const a = document.createElement("a");
  a.href = url;
  a.download = filename;

  // Append the anchor to the body (required for Firefox)
  document.body.appendChild(a);

  // Trigger the download by simulating a click
  a.click();

  // Clean up: remove the anchor and revoke the Blob URL
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}
