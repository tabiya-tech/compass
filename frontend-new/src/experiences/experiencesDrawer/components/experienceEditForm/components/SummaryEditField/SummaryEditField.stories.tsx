import type { Meta, StoryObj } from "@storybook/react";
import { action } from "@storybook/addon-actions";

import SummaryEditField from "src/experiences/experiencesDrawer/components/experienceEditForm/components/SummaryEditField/SummaryEditField";
import UserPreferencesStateService from "src/userPreferences/UserPreferencesStateService";
import ExperienceService from "src/experiences/experienceService/experienceService";

const meta: Meta<typeof SummaryEditField> = {
  title: "Experiences/ExperienceEditForm/SummaryEditField",
  component: SummaryEditField,
  tags: ["autodocs"],
  decorators: [
    (Story) => {
      // Mock AuthenticationStateService
      const mockUserPreferencesStateService = UserPreferencesStateService.getInstance();
      mockUserPreferencesStateService.getActiveSessionId = () => 1234;

      const mockExperienceService = {
        getOriginalExperience: () => ({
          uuid: "1234",
          summary: "This is the original summary of the experience.".repeat(15),
        }),
        updateExperience: (sessionId: number, experience_uuid: string, data: { summary: string }) => {
          return new Promise((resolve) => {
            setTimeout(() => {
              resolve({
                ...data,
                uuid: experience_uuid,
              });
            }, 1000);
          });
        },
      };
      // @ts-ignore
      ExperienceService.getInstance().getUneditedExperience = mockExperienceService.getOriginalExperience;
      // @ts-ignore
      ExperienceService.getInstance().updateExperience = mockExperienceService.updateExperience;

      return (
        <div style={{ width: "40%", padding: "20px" }}>
          <Story />
        </div>
      );
    },
  ],
};

export default meta;
type Story = StoryObj<typeof SummaryEditField>;

export const Shown: Story = {
  args: {
    summary: "This is a summary of the experience.",
    notifyOnChange: action("notifyOnChange"),
  },
};

export const ShownWithLongText: Story = {
  args: {
    summary: "This is a summary of the experience. ".repeat(15),
    notifyOnChange: action("notifyOnChange"),
  },
};

export const Empty: Story = {
  args: {
    summary: "",
    notifyOnChange: action("notifyOnChange"),
  },
};

export const Error: Story = {
  args: {
    summary: "This is a summary of the experience.",
    notifyOnChange: action("notifyOnChange"),
    error: "An error has occurred",
  },
};

export const Disabled: Story = {
  args: {
    summary: "This is a summary of the experience.",
    notifyOnChange: action("notifyOnChange"),
  },
};
