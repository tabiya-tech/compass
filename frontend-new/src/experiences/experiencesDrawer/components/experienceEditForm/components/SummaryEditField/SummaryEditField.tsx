import { useContext, useState, useEffect } from "react";
import {
  Experience,
  SUMMARY_MAX_LENGTH
} from "src/experiences/experienceService/experiences.types";
import { Box, Stack, Typography, useTheme } from "@mui/material";
import InlineEditField from "src/theme/InlineEditField/InlineEditField";
import RestoreIcon from "@mui/icons-material/Restore";
import { IsOnlineContext } from "src/app/isOnlineProvider/IsOnlineProvider";
import UserPreferencesStateService from "src/userPreferences/UserPreferencesStateService";
import ExperienceService from "src/experiences/experienceService/experienceService";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";
import { ExperienceError } from "src/error/commonErrors";
import CustomLink from "src/theme/CustomLink/CustomLink";

export interface SummaryEditFieldProps {
  summary: string;
  experience_uuid: string;
  notifyOnChange: (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>, field: keyof Experience, maxLength?: number) => void;
  error?: string;
}

const uniqueId = "b1a6b2cd-cfac-46ff-9c35-0298d87def6b";

export const DATA_TEST_ID = {
  FORM_SUMMARY: `form-summary-${uniqueId}`,
  FORM_SUMMARY_RESTORE: `form-summary-restore-${uniqueId}`,
  FORM_SUMMARY_HELPER: `form-summary-error-${uniqueId}`,
};

const StyledCustomLink: React.FC<React.ComponentProps<typeof CustomLink>> = (props) => {
  const theme = useTheme();
  return (
    <CustomLink
      {...props}
      style={{
        display: "inline-flex",
        flexDirection: "row",
        gap: theme.spacing(theme.tabiyaSpacing.xs),
        verticalAlign: "bottom",
        ...props.style
      }}
    />
  );
};

export const SUMMARY_FIELD_NAME : keyof Experience = "summary";

const SummaryEditField: React.FC<Readonly<SummaryEditFieldProps>> = ({
  summary,
  experience_uuid,
  notifyOnChange,
  error,
}) => {
  const theme = useTheme();
  const isOnline = useContext(IsOnlineContext)
  const { enqueueSnackbar } = useSnackbar();


  const [summaryValue, setSummaryValue] = useState(summary);
  const [isRestoring, setIsRestoring] = useState(false);
  const [errorText, setErrorText] = useState<string | null>(error ?? null);
  const [successText, setSuccessText] = useState<string | null>(null);

  const restoreToOriginalSummary = async () => {
      setIsRestoring(true);
      const sessionId = UserPreferencesStateService.getInstance().getActiveSessionId();
      if (!sessionId) {
        throw new Error("User has no sessions");
      }

      try {
        // get original version of summary
        const experienceService = ExperienceService.getInstance();
        const originalExperience = await experienceService.getUneditedExperience(sessionId, experience_uuid);
        setSummaryValue(originalExperience.summary ?? "");
        setSuccessText("Summary restored.")
        notifyOnChange({ target: { value: originalExperience.summary ?? "" } } as React.ChangeEvent<HTMLTextAreaElement>, SUMMARY_FIELD_NAME, SUMMARY_MAX_LENGTH);
        setTimeout(() => {
          setSuccessText(null);
        }, 3000); // Clear message after 3 seconds
      } catch (err) {
        console.error(new ExperienceError("Failed to restore summary:", err));
        enqueueSnackbar("Failed to restore summary. Please try again later.", { variant: "error" });
      } finally {
        setIsRestoring(false);
      }
  }

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      setSummaryValue(event.target.value);
      notifyOnChange(event, SUMMARY_FIELD_NAME, SUMMARY_MAX_LENGTH);
  }
  // decide when to show error messages
  useEffect(() => {
    if (error) {
      setErrorText(error);
    } else {
      setErrorText(null);
    }
  }, [error, theme.palette.error.main]);


  return (
    <Box>
      <Box
        sx={{
          backgroundColor: theme.palette.grey[100],
          borderRadius: theme.fixedSpacing(theme.tabiyaRounding.sm),
          width: "100%",
        }}
      >
        {/* Toolbar */}
        <Box
          sx={{
            display: "flex",
            alignItems: "center",
            justifyContent: "space-between",
            borderBottom: `1px solid ${theme.palette.grey[200]}`,
            padding: theme.fixedSpacing(theme.tabiyaSpacing.sm),
          }}
        >
          <Stack direction="row" spacing={1}>
            <StyledCustomLink
              onClick={restoreToOriginalSummary}
              disabled={!isOnline || isRestoring}
              data-testid={DATA_TEST_ID.FORM_SUMMARY_RESTORE}
              variant="caption"
            >
              <RestoreIcon /> Restore
            </StyledCustomLink>
            {/* Other buttons */}
          </Stack>
        </Box>

        {/* Text area */}
        <Box
          sx={{
            padding: theme.spacing(theme.tabiyaSpacing.sm),
          }}
        >
          <InlineEditField
            placeholder="Summary of experience"
            value={summaryValue}
            onChange={handleInputChange}
            multiline
            minRows={8}
            inputProps={{ maxLength: SUMMARY_MAX_LENGTH + 1 }} // +1 to allow the user to overflow and see the error
            sx={{
              border: 0,
              padding: theme.fixedSpacing(theme.tabiyaSpacing.sm),
              fontSize: theme.typography.body2.fontSize,
              backgroundColor: "transparent",
              "& textarea": {
                padding: 0,
                resize: "none",
              },
            }}
            data-testid={DATA_TEST_ID.FORM_SUMMARY}
            error={!!error}
          />
        </Box>
      </Box>
    {( errorText || successText) && (
      <Typography
        variant="caption"
        color={errorText ? theme.palette.error.main : theme.palette.success.dark}
        data-testid={DATA_TEST_ID.FORM_SUMMARY_HELPER}
      >
        {errorText ?? successText}
      </Typography>
    )}
  </Box>
  );
};

export default SummaryEditField;
