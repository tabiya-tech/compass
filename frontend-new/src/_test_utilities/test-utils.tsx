// Based on https://testing-library.com/docs/react-testing-library/setup/
import React, { ReactElement, ReactNode } from "react";
import { render, renderHook, RenderHookOptions, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";

// Import the Firebase mock utilities
import "src/_test_utilities/firebaseMock";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/isOnlineProvider/IsOnlineProvider";
jest.mock("firebase/compat/app", () => {
  return {
    initializeApp: jest.fn(),
    auth: jest.fn().mockReturnValue({
      signOut: jest.fn(),
      currentUser: {
        getIdToken: jest.fn(),
        getTokenResult: jest.fn(),
      },
      onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
      signInWithPopup: jest.fn(),
      GoogleAuthProvider: { PROVIDER_ID: "google.com" },
    }),
  };
});

jest.mock("firebaseui", () => {
  return {
    auth: {
      AuthUI: {
        start: jest.fn(),
        getInstance: jest.fn().mockReturnValue({
          start: jest.fn(),
          reset: jest.fn(),
        }),
        reset: jest.fn(),
      },
    },
  };
});

jest.mock("@sentry/react", () => ({
  withSentry: (Component: any) => Component,
  getFeedback: jest.fn(),
  withProfiler: (Component: any) => Component,
  wrapCreateBrowserRouter: (Component: any) => Component,
}));

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

function customRenderHook<T>(
  hook: () => T,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { wrapper?: React.ComponentType }
) {
  const CombinedWrapper = ({ children }: { children: ReactNode }) => {
    const Wrapper = options?.wrapper ?? React.Fragment;
    return (
      <AllTheProviders>
        <Wrapper>{children}</Wrapper>
      </AllTheProviders>
    );
  };

  return renderHook(hook, { wrapper: CombinedWrapper, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
export { customRenderHook as renderHook };
