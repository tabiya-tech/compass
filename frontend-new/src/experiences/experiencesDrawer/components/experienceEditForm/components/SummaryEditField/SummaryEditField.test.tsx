import "src/_test_utilities/consoleMock";
import React from "react";
import { render, screen, fireEvent, waitFor, act, within } from "src/_test_utilities/test-utils";
import userEvent from "@testing-library/user-event";
import SummaryEditField, { DATA_TEST_ID } from "src/experiences/experiencesDrawer/components/experienceEditForm/components/SummaryEditField/SummaryEditField";
import ExperienceService from "src/experiences/experienceService/experienceService";
import UserPreferencesStateService from "src/userPreferences/UserPreferencesStateService";
import { ExperienceError } from "src/error/commonErrors";
import { Experience, SUMMARY_MAX_LENGTH } from "src/experiences/experienceService/experiences.types";
import { mockBrowserIsOnLine } from "src/_test_utilities/mockBrowserIsOnline";
import { resetAllMethodMocks } from "src/_test_utilities/resetAllMethodMocks";

jest.mock("src/theme/SnackbarProvider/SnackbarProvider", () => {
  const actual = jest.requireActual("src/theme/SnackbarProvider/SnackbarProvider");
  return {
    ...actual,
    __esModule: true,
    useSnackbar: jest.fn().mockReturnValue({
      enqueueSnackbar: jest.fn(),
      closeSnackbar: jest.fn(),
    }),
  };
});

const mockNotifyOnChange = jest.fn();
const mockEnqueueSnackbar = require("src/theme/SnackbarProvider/SnackbarProvider").useSnackbar().enqueueSnackbar;

const defaultProps = {
  summary: "This is a summary of the experience.",
  experience_uuid: "exp-uuid-123",
  notifyOnChange: mockNotifyOnChange,
};

describe("SummaryEditField", () => {
  beforeEach(() => {
    mockBrowserIsOnLine(true);
    jest.clearAllMocks();
    // Mock session id
    UserPreferencesStateService.getInstance().setUserPreferences({
      sessions: [1234],
      user_feedback_answered_questions: {},
    } as any);
    resetAllMethodMocks(UserPreferencesStateService.getInstance())
  });

  test("should render with initial summary and no error", () => {
    // GIVEN default props
    render(<SummaryEditField {...defaultProps} />);
    // THEN summary field is present and has correct value
    const summaryContainer = screen.getByTestId(DATA_TEST_ID.FORM_SUMMARY);
    const textarea = within(summaryContainer).getByRole("textbox");
    expect(textarea).toBeInTheDocument();
    expect(textarea).toHaveValue(defaultProps.summary);

    // AND restore button is present and enabled
    const restoreBtn = screen.getByTestId(DATA_TEST_ID.FORM_SUMMARY_RESTORE);
    expect(restoreBtn).toBeInTheDocument();
    expect(restoreBtn).not.toBeDisabled();
    // AND no error or success message is shown
    expect(screen.queryByTestId(DATA_TEST_ID.FORM_SUMMARY_HELPER)).not.toBeInTheDocument();
    // AND it should match the snapshot
    expect(screen.getByTestId(DATA_TEST_ID.FORM_SUMMARY)).toMatchSnapshot();
  });

  test("should render with error message if error prop is provided", () => {
    // GIVEN error prop
    render(<SummaryEditField {...defaultProps} error="Some error" />);
    // THEN error message is shown
    expect(screen.getByTestId(DATA_TEST_ID.FORM_SUMMARY_HELPER)).toHaveTextContent("Some error");
  });

  test("should call notifyOnChange and update value when summary is changed", () => {
    // GIVEN rendered component
    render(<SummaryEditField {...defaultProps} />);
    const summaryContainer = screen.getByTestId(DATA_TEST_ID.FORM_SUMMARY);
    const textarea = within(summaryContainer).getByRole("textbox");
    // THEN textarea is present
    expect(textarea).toBeInTheDocument();
    // WHEN user types in textarea
    userEvent.clear(textarea!);
    fireEvent.change(textarea!, { target: { value: "New summary" } });
    // THEN notifyOnChange is called
    expect(mockNotifyOnChange).toHaveBeenCalled();
    // AND value is updated
    expect(textarea).toHaveValue("New summary");
  });

  test("should restore summary to original when restore is clicked (success path)", async () => {
    jest.useFakeTimers()
    // GIVEN original summary from service
    const originalSummary = "Original summary from service.";
    jest.spyOn(ExperienceService.getInstance(), "getOriginalExperience").mockResolvedValue({ summary: originalSummary } as Experience);
    render(<SummaryEditField {...defaultProps} />);
    // WHEN restore is clicked
    const restoreBtn = screen.getByTestId(DATA_TEST_ID.FORM_SUMMARY_RESTORE);
    userEvent.click(restoreBtn);
    // THEN summary is updated
    await waitFor(() => {
      const summaryContainer = screen.getByTestId(DATA_TEST_ID.FORM_SUMMARY);
      expect(within(summaryContainer).getByRole("textbox")).toHaveValue(originalSummary);
    });
    // AND success message is shown
    expect(screen.getByTestId(DATA_TEST_ID.FORM_SUMMARY_HELPER)).toHaveTextContent("Summary restored to original version.");
    // AND notifyOnChange is called with original summary
    expect(mockNotifyOnChange).toHaveBeenCalledWith(
      expect.objectContaining({ target: { value: originalSummary } }),
      "summary",
      SUMMARY_MAX_LENGTH
    );
    // AND success message disappears after timeout
    await act(async () => {
      jest.advanceTimersByTime(3010);
    });
    expect(screen.queryByTestId(DATA_TEST_ID.FORM_SUMMARY_HELPER)).not.toBeInTheDocument();
    // cleanup
    jest.useRealTimers();
  });

  test("should show error snackbar and log error if restore fails", async () => {
    // GIVEN service throws error
    const givenError = new Error("some error");
    jest.spyOn(ExperienceService.getInstance(), "getOriginalExperience").mockRejectedValueOnce(givenError);
    render(<SummaryEditField {...defaultProps} />);
    // WHEN restore is clicked
    const restoreBtn = screen.getByTestId(DATA_TEST_ID.FORM_SUMMARY_RESTORE);
    userEvent.click(restoreBtn);
    // THEN error is logged and snackbar is shown
    await waitFor(() => {
      expect(console.error).toHaveBeenCalledWith(new ExperienceError("Failed to restore original summary:", givenError));
    });
    expect(mockEnqueueSnackbar).toHaveBeenCalledWith(
      "Failed to restore summary. Please try again later.",
      { variant: "error" }
    );
    // AND the notifyOnChange function is not called
    expect(mockNotifyOnChange).not.toHaveBeenCalled();
  });

  test("should disable restore button while restoring", async () => {
    // GIVEN slow service
    let resolvePromise: any;
    jest.spyOn(ExperienceService.getInstance(), "getOriginalExperience").mockReturnValueOnce(
      new Promise((res) => { resolvePromise = res; })
    );
    jest.spyOn(ExperienceService.getInstance(), "updateExperience").mockResolvedValue({ summary: "restored" } as Experience);
    // WHEN component is rendered
    render(<SummaryEditField {...defaultProps} />);
    const restoreBtn = screen.getByTestId(DATA_TEST_ID.FORM_SUMMARY_RESTORE);
    // WHEN restore is clicked
    userEvent.click(restoreBtn);
    // THEN button is disabled
    await waitFor(() => {
      expect(restoreBtn).toHaveAttribute("aria-disabled", "true");
    });
    // WHEN promise resolves
    act(() => {
      resolvePromise({ summary: "restored" });
    });
    // THEN button is enabled again
    await waitFor(() => expect(restoreBtn).not.toBeDisabled());
  });

  test("should disable restore button if offline", async () => {
    // GIVEN offline context
    mockBrowserIsOnLine(false)
    render(<SummaryEditField {...defaultProps} />);
    // THEN restore button is disabled
    await waitFor(() => {
      expect(screen.getByTestId(DATA_TEST_ID.FORM_SUMMARY_RESTORE)).toHaveAttribute("aria-disabled", "true");
    });
  });
}); 