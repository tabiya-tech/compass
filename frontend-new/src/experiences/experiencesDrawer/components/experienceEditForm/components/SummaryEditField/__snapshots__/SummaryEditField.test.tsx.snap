// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryEditField should render with initial summary and no error 1`] = `
<div
  class="MuiFormControl-root MuiTextField-root css-1tbudcr-MuiFormControl-root-MuiTextField-root"
  data-testid="form-summary-b1a6b2cd-cfac-46ff-9c35-0298d87def6b"
>
  <div
    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-1fv69vb-MuiInputBase-root-MuiOutlinedInput-root"
  >
    <textarea
      aria-invalid="false"
      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-cmpe59-MuiInputBase-input-MuiOutlinedInput-input"
      id=":r0:"
      maxlength="1001"
      placeholder="Summary of experience"
      style="height: 0px; overflow: hidden;"
    >
      This is a summary of the experience.
    </textarea>
    <textarea
      aria-hidden="true"
      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-cmpe59-MuiInputBase-input-MuiOutlinedInput-input"
      readonly=""
      style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
      tabindex="-1"
    />
    <fieldset
      aria-hidden="true"
      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        class="css-ihdtdm"
      >
        <span
          class="notranslate"
        >
          â€‹
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
